File: public-structures/src/main/java/top/yudoge/pojos/ResourceEntry.java
package top.yudoge.pojos;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import top.yudoge.converter.ResourceEntryDeserializer;

import java.util.Iterator;
import java.util.List;

@JsonDeserialize(using = ResourceEntryDeserializer.class)
public interface ResourceEntry {
    Boolean isDirectory();
    String getName();
    void setName(String name);
    Long getSize();
    void setSize(Long size);
    List<ResourceEntry> getChildren();
    void addChild(ResourceEntry child);
}

File: pyscripts/resource_spider/resource_entry.py
class ResourceEntry(dict):
    def __init__(self, name, size, directory, _class):
        self['name'] = name
        self['size'] = size
        self['directory'] = directory
        self['_class'] = _class

class File(ResourceEntry):
    def __init__(self, name, size):
        ResourceEntry.__init__(self, name, size, False, 'top.yudoge.pojos.File')

class Directory(ResourceEntry):
    def __init__(self, name):
        ResourceEntry.__init__(self, name, 0, True, 'top.yudoge.pojos.Directory')
        self['children'] = []

    def addChild(self, entry):
        self['children'].append(entry)
        self['size'] += entry['size']